lib/TH/THMemoryFile.c

```diff
@@ -383,12 +384,12 @@ static size_t THMemoryFile_readLong(THFile *self, long *data, size_t n)
     {
       int i, big_endian = !THDiskFile_isLittleEndianCPU();
       size_t nByte = 8*n;
-      long *storage = (long *)(mfself->storage->data + mfself->position);
+      int32_t *storage = (int32_t *)(mfself->storage->data + mfself->position);
```

Comment:
It should be int32_t. This else branch is to handle sizeof(long)!=8 while longSize==8 case (for example type long is of 4 bytes). Current implementation has a trick to use two 4bytes integers to represent one 8bytes long. Please check the main implementation in the following lines.
PS. I do not think this case is useful as long as the input "data" is of "long *". Do you know an use case for this?